src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,629 @@
+   // field path. The second element is the value.
+   if (!bson_iter_next (&array)) {
+      CLIENT_ERR (
+         "expected 2 elements in operand %s, got 0: %s", op_type_str, str);
+      goto fail;
+   }
+   if (!BSON_ITER_HOLDS_UTF8 (&array)) {
+      CLIENT_ERR (
+         "expected UTF-8 as first element in %s: %s", op_type_str, str);
+      goto fail;
+   }
+   field = bson_iter_utf8 (&array, NULL);
+
+   if (!bson_iter_next (&array)) {
+      CLIENT_ERR (
+         "expected 2 elements in operand %s, got 1: %s", op_type_str, str);
+      goto fail;
+   }
+   value = array;
+   if (bson_iter_next (&array)) {
+      CLIENT_ERR (
+         "expected 2 elements in operand %s, got > 2: %s", op_type_str, str);
+      goto fail;
+   }
+
+   out->field = field;
+   out->op_type_str = op_type_str;
+   out->op_type = get_operator_type (op_type_str);
+   out->value = value;
+   ok = true;
+fail:
+   return ok;
+}
+
+// Parses a document like {age: {$gt: 5}}.
+static bool
+parse_match_expression (bson_iter_t in,
+                        const char *str,
+                        operator_value_t *out,
+                        mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (str);
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bson_iter_t document, value;
+   const char *op_type_str;
+   bool ok = false;
+   const char *field = bson_iter_key (&in);
+
+   if (!BSON_ITER_HOLDS_DOCUMENT (&in)) {
+      CLIENT_ERR ("expected argument to be document: %s", str);
+      goto fail;
+   }
+
+   if (!bson_iter_recurse (&in, &document)) {
+      CLIENT_ERR ("failed to recurse into document: %s", str);
+      goto fail;
+   }
+
+   // Expect exactly 1 elements, like {$gt: 5}.
+   if (!bson_iter_next (&document)) {
+      CLIENT_ERR ("expected 1 elements in operand %s, got 0: %s", field, str);
+      goto fail;
+   }
+   op_type_str = bson_iter_key (&document);
+   if (!is_supported_operator (op_type_str)) {
+      CLIENT_ERR ("unsupported operator: %s", op_type_str);
+      goto fail;
+   }
+   value = document;
+
+   if (bson_iter_next (&document)) {
+      CLIENT_ERR ("expected 1 elements in operand %s, got > 1: %s", field, str);
+      goto fail;
+   }
+
+   out->field = field;
+   out->op_type_str = op_type_str;
+   out->op_type = get_operator_type (op_type_str);
+   out->value = value;
+
+   ok = true;
+fail:
+   return ok;
+}
+
+bool
+mc_FLE2RangeFindDriverSpec_parse (mc_FLE2RangeFindDriverSpec_t *spec,
+                                  const bson_t *in,
+                                  mongocrypt_status_t *status)
+{
+   *spec = (mc_FLE2RangeFindDriverSpec_t){0};
+   // `in` may be an Aggregate Expression with this form:
+   // {$and: [{$gt: ["$age", 5]}, {$lt:["$age", 50]}]}
+   // Or `in` may be a Match Expression with this form:
+   // {$and: [{age: {$gt: 5}}, {age: {$lt: 50}} ]}
+   bson_iter_t and, array;
+   bool ok = false;
+   char *str = bson_as_canonical_extended_json (in, NULL);
```

Comment:
IIUC, `str` is only used in error messages in the `parse_` functions. Recommend passing the `bson_t` directly to these functions and have them generate the extended JSON string only when they need it in error messages.

Reply:
Good idea. Added a `ERR_WITH_BSON` macro to defer creating JSON string